### PR TITLE
Add micro-benchmark for Ray and Frustum intersections

### PR DESCRIPTION
**Pull request by paulgessinger**
Previously I had only benchmarked this statistically. This PR adds benchmarks based on @HadrienG2's tooling. 

- For Rays, it uses a single bounding box and randomly selected rays. The performance should not at all depend on the concrete ray parameters
- For Frustums, the performances does in fact depend on the orientation. The bench includes 6 extreme cases which establish what I believe to be a best-case and a worst-case performance. I only test 4 sided frustums here.

Also included are a bunch of alternative implementation of the intersection algorithms, so have something to compare against. They are mostly slight variations of the main implementation. [I can remove this if we don't care about these]

Output should be something like

```console

==== RAY ====

Make sure ray implementations are identical
Seems ok
Benchmarking rays nominal: 20000 runs of 10000 iteration(s), 1064.0ms total, 53.0890+/-0.0013µs per run, 5.309+/-0.013ns per iteration
Benchmarking rays naive 1: 20000 runs of 10000 iteration(s), 2583.0ms total, 128.6740+/-0.0017µs per run, 12.867+/-0.017ns per iteration
Benchmarking rays naive 2: 20000 runs of 10000 iteration(s), 2578.7ms total, 128.3140+/-0.0010µs per run, 12.831+/-0.010ns per iteration
Benchmarking rays naive 3: 20000 runs of 10000 iteration(s), 2602.6ms total, 128.8770+/-0.0010µs per run, 12.888+/-0.010ns per iteration
Benchmarking rays naive 4: 20000 runs of 10000 iteration(s), 1488.4ms total, 73.8220+/-0.0018µs per run, 7.382+/-0.018ns per iteration
Benchmarking rays naive 5: 20000 runs of 10000 iteration(s), 1651.8ms total, 81.3520+/-0.0032µs per run, 8.135+/-0.032ns per iteration

==== FRUSTUM ====

Make sure frustum implementations are identical
Seems ok
------------------------
away
------------------------
Benchmarking frust nominal: 20000 runs of 1000 iteration(s), 276.3ms total, 13.6580+/-0.0003µs per run, 13.658+/-0.009ns per iteration
Benchmarking frust opt 1: 20000 runs of 1000 iteration(s), 275.6ms total, 13.7130+/-0.0008µs per run, 13.713+/-0.026ns per iteration
Benchmarking frust opt 2: 20000 runs of 1000 iteration(s), 254.0ms total, 12.5400+/-0.0002µs per run, 12.540+/-0.006ns per iteration
Benchmarking frust opt 3: 20000 runs of 1000 iteration(s), 848.3ms total, 42.2470+/-0.0020µs per run, 42.247+/-0.064ns per iteration
Benchmarking frust opt 4: 20000 runs of 1000 iteration(s), 863.3ms total, 43.0400+/-0.0012µs per run, 43.040+/-0.038ns per iteration
------------------------
towards
------------------------
Benchmarking frust nominal: 20000 runs of 1000 iteration(s), 978.1ms total, 48.7990+/-0.0005µs per run, 48.799+/-0.015ns per iteration
Benchmarking frust opt 1: 20000 runs of 1000 iteration(s), 977.7ms total, 48.6440+/-0.0005µs per run, 48.644+/-0.015ns per iteration
Benchmarking frust opt 2: 20000 runs of 1000 iteration(s), 944.4ms total, 47.0260+/-0.0008µs per run, 47.026+/-0.024ns per iteration
Benchmarking frust opt 3: 20000 runs of 1000 iteration(s), 981.8ms total, 48.9160+/-0.0004µs per run, 48.916+/-0.014ns per iteration
Benchmarking frust opt 4: 20000 runs of 1000 iteration(s), 962.6ms total, 47.9380+/-0.0006µs per run, 47.938+/-0.018ns per iteration
------------------------
left
------------------------
Benchmarking frust nominal: 20000 runs of 1000 iteration(s), 980.6ms total, 48.7720+/-0.0006µs per run, 48.772+/-0.019ns per iteration
Benchmarking frust opt 1: 20000 runs of 1000 iteration(s), 975.4ms total, 48.6660+/-0.0011µs per run, 48.666+/-0.033ns per iteration
Benchmarking frust opt 2: 20000 runs of 1000 iteration(s), 950.3ms total, 47.1390+/-0.0009µs per run, 47.139+/-0.030ns per iteration
Benchmarking frust opt 3: 20000 runs of 1000 iteration(s), 985.5ms total, 48.9580+/-0.0007µs per run, 48.958+/-0.022ns per iteration
Benchmarking frust opt 4: 20000 runs of 1000 iteration(s), 969.1ms total, 48.3010+/-0.0008µs per run, 48.301+/-0.024ns per iteration
------------------------
right
------------------------
Benchmarking frust nominal: 20000 runs of 1000 iteration(s), 632.1ms total, 31.5190+/-0.0006µs per run, 31.519+/-0.018ns per iteration
Benchmarking frust opt 1: 20000 runs of 1000 iteration(s), 632.7ms total, 31.4380+/-0.0008µs per run, 31.438+/-0.025ns per iteration
Benchmarking frust opt 2: 20000 runs of 1000 iteration(s), 621.8ms total, 30.7740+/-0.0005µs per run, 30.774+/-0.015ns per iteration
Benchmarking frust opt 3: 20000 runs of 1000 iteration(s), 919.7ms total, 45.8850+/-0.0007µs per run, 45.885+/-0.022ns per iteration
Benchmarking frust opt 4: 20000 runs of 1000 iteration(s), 918.3ms total, 45.7580+/-0.0010µs per run, 45.758+/-0.031ns per iteration
------------------------
up
------------------------
Benchmarking frust nominal: 20000 runs of 1000 iteration(s), 802.4ms total, 40.0470+/-0.0007µs per run, 40.047+/-0.023ns per iteration
Benchmarking frust opt 1: 20000 runs of 1000 iteration(s), 803.3ms total, 40.0370+/-0.0011µs per run, 40.037+/-0.036ns per iteration
Benchmarking frust opt 2: 20000 runs of 1000 iteration(s), 777.4ms total, 38.7060+/-0.0005µs per run, 38.706+/-0.016ns per iteration
Benchmarking frust opt 3: 20000 runs of 1000 iteration(s), 956.4ms total, 47.6680+/-0.0004µs per run, 47.668+/-0.014ns per iteration
Benchmarking frust opt 4: 20000 runs of 1000 iteration(s), 944.4ms total, 47.0900+/-0.0039µs per run, 47.090+/-0.124ns per iteration
------------------------
down
------------------------
Benchmarking frust nominal: 20000 runs of 1000 iteration(s), 457.5ms total, 22.7410+/-0.0003µs per run, 22.741+/-0.011ns per iteration
Benchmarking frust opt 1: 20000 runs of 1000 iteration(s), 454.1ms total, 22.6220+/-0.0005µs per run, 22.622+/-0.015ns per iteration
Benchmarking frust opt 2: 20000 runs of 1000 iteration(s), 438.4ms total, 21.8670+/-0.0009µs per run, 21.867+/-0.029ns per iteration
Benchmarking frust opt 3: 20000 runs of 1000 iteration(s), 885.6ms total, 44.0970+/-0.0006µs per run, 44.097+/-0.018ns per iteration
Benchmarking frust opt 4: 20000 runs of 1000 iteration(s), 882.3ms total, 43.9890+/-0.0010µs per run, 43.989+/-0.030ns per iteration
```

### CODE REVIEW

File: Tests/Benchmarks/RayFrustumBenchmark.cpp [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+  vertex_array_type t1s = (box.min() - origin).array() * idir;
+
+  vertex_array_type tsmaller = t0s.min(t1s);
+  vertex_array_type tbigger = t0s.max(t1s);
+
+  value_type tmin = tsmaller.maxCoeff();
+  value_type tmax = tbigger.minCoeff();
+
+  return tmin < tmax && tmax > 0.0;
+}
```
Comment (HadrienG2) [inline Tests/Benchmarks/RayFrustumBenchmark.cpp:47]: This is only used for results comparison with naiveIntersect. Maybe a development artifact from an earlier version of this microbenchmark?

**Comment by paulgessinger** [comment]
> * Why are there so many different versions of each calculation? What purpose does it serve?

I wanted to compare a variety of versions to determine what the impact of different "optimizations" is. I can remove this if we don't care, but I think it's helpful.

> * Why are almost all intersection computations "naive" and almost all frustum operations "opt"?

The names are a bit arbitrary. I can rename them to just be `frustumVariantN`.

> * The functions that are used for benchmarking are listed twice, once for validation and once for proper benchmarking, and as we've seen the two lists can go out of sync. Would it be possible to avoid this problem by listing the functions only once?   
>   * As a possible roadblock, I think avoiding indirect function call overhead while doing this may require a bit of `std::tuple` template metaprogramming.

Yeah I can maybe list them only once.

File: Tests/Benchmarks/RayFrustumBenchmark.cpp (hunk 1 above)
Comment (paulgessinger) [reply Tests/Benchmarks/RayFrustumBenchmark.cpp:47]: Yeah I think this isn't used. True

**Comment by paulgessinger** [comment]
I'll clean this up a bit.

**Comment by paulgessinger** [comment]
I tried to improve the benchmark here, removed a few redundancies, and improved labelling. Could you have another look @HadrienG2?

**Comment by HadrienG2** [comment]
That certainly looks a lot nicer! Did you check that the indirect function call doesn't affect your results too much?

**Comment by paulgessinger** [comment]
> That certainly looks a lot nicer! Did you check that the indirect function call doesn't affect your results too much?

You mean me calling the 'nominal' implementation from a lambda?

**Comment by HadrienG2** [comment]
@paulgessinger Yes, I don't think the compiler can remove the function pointer indirection when it is accessed through an std::map.

**Comment by paulgessinger** [comment]
But the map access only happens once, and the wrapping lambda should more or less be compiled away, no? I think that's also what I'm seeing in the numbers

**Comment by HadrienG2** [comment]
Maybe we're actually not talking about the same thing. I'm using a cellphone here, which gives a bit of a limited perspective...

As far as I can see, you pass to microBenchmark a simple lambda, which will indeed almost certainly be inlined. But that lambda, in turn, calls a function pointer which represents the particular intersection algorithm that is being tested. This function pointer was obtained from an std::map, which AFAIK is too complex for a compiler to const-propagate through. So I assume that the compiler will not be able to trace it back to the original map entry insertion, and replace it with the benchmark's code. Therefore, this std::map based version should feature a function call which the original did not have.

**Comment by paulgessinger** [comment]
Hm, just comparing my current measurements with the `map` to what I posted in the original PR description, I don't see a slow down. I don't get the exact same numbers, but the ones I get now don't seem to be worse.

Do you think that's good enough for now?

**Comment by HadrienG2** [comment]
If the numbers are similar enough, then let's get this in as is :)

**Pull request merged**